BlockDataSource/Classes/DataSource.swift

```diff
@@ -0,0 +1,598 @@
+            self.header = header
+            self.items = items
+            self.footer = footer
+        }
+        
+        /// Convenience init for a section with a single item
+        public init(header: HeaderFooter? = nil, item: DataSourceItem, footer: HeaderFooter? = nil) {
+            self.header = header
+            self.items = [item]
+            self.footer = footer
+        }
+        
+        // Reference items with `section[index]`
+        public subscript(index: Int) -> DataSourceItem {
+            return items[index]
+        }
+    }
+
+
+    // MARK: -
+
+    public struct ListItem: DataSourceItem {
+        
+        // The block which configures the cell
+        public var configure: ConfigureBlock
+
+        // The block that executes when the cell is tapped
+        public var onSelect: IndexPathBlock?
+        
+        // The block that executes when the cell is deleted
+        public var onDelete: IndexPathBlock?
+        
+        // Lets the dataSource know that this item can be reordered
+        public var reorderable: Bool = false
+
+        public var viewClass: UIView.Type
+
+        private var customReuseIdentifier: String?
+
+        public var reuseIdentifier: String {
+            if let customReuseIdentifier = customReuseIdentifier {
+                return customReuseIdentifier
+            } else {
+                return String(describing: viewClass)
+            }
+        }
+
+        /**
+         Initialize a item
+         
+           - parameters:
+             - configure: The cell configuration block.
+             - onSelect: The closure to execute when the cell is tapped
+             - onDelete: The closure to execute when the cell is deleted
+             - reorderable: Flag to indicate if this cell can be reordered
+             - customReuseIdentifier: Set to override the default reuseIdentifier. Default is nil.
+         */
+        public init<Cell: UITableViewCell>(configure: @escaping (Cell) -> Void, onSelect: IndexPathBlock? = nil, onDelete: IndexPathBlock? = nil, reorderable: Bool = true, customReuseIdentifier: String? = nil) {
+            self.onSelect = onSelect
+            self.onDelete = onDelete
+            self.reorderable = reorderable
+            self.customReuseIdentifier = customReuseIdentifier
+            
+            self.viewClass = Cell.self
+            self.configure = { cell in
+                configure(cell as! Cell)
+            }
+        }
+        
+        // Convienence init for trailing closure syntax
+        public init<Cell: UITableViewCell>(reorderable: Bool = true, configure: @escaping (Cell) -> Void) {
+            self.init(
+                configure: configure,
+                onSelect: nil,
+                onDelete: nil,
+                reorderable: reorderable
+            )
+        }
+    }
+
+
+    // MARK: -
+
+    public struct GridItem: DataSourceItem {
```

Comment:
I want to enforce that you can only construct a DataSource using a specific kind of `DataSourceItem` to prevent people from constructing datasources with both list and grid items. Could be easy to cause runtime crashes.